text/1004-transactions-api.rst

```diff
@@ -0,0 +1,619 @@
+
+Add ``RetryOptions`` class:
+
+.. code-block:: Python
+
+    type DelayFn = (attempt: number) => number;
+
+    class AttempsOption(enum.Enum):
+        ALL = "ALL"
+        NETWORK_ERROR = "NETWORK_ERROR"
+        CONCURRENT_UPDATE = "CONCURRENT_UPDATE"
+        DEADLOCK = "DEADLOCK"
+
+    class RetryOptions:
+        def __init__(self, attempts=3): ...
+
+        def transaction_options(self,
+           isolation: str = None,
+           readonly: bool = None,
+           deferrable: bool = None,
+        ) -> RetryOptions: ...
+
+        def attempts(
+            which: AttemptsOption,
+            attempts: int,
+            delay_ms: Callable[[int], [float]],
+        ): RetryOptions: ...
+    }
+
+Introduce the abstract classes for queries:
+
+.. code-block:: python
+
+    class AsyncExecutor(abc.AbstraceBaseClass):
+        async def execute(self, query): ...
+        async def query(self, query: str, *args, **kwargs) -> datatypes.Set: ...
+        async def query_one(self, query: str, *args, **kwargs) -> typing.Any: ...
+        async def query_json(self, query: str, *args, **kwargs) -> str: ...
+        async def query_one_json(self, query: str, *args, **kwargs) -> str: ...
+
+    class Executor(abc.AbstractClass):
+        def query(self, query: str, *args, **kwargs) -> datatypes.Set: ...
+        def query_one(self, query: str, *args, **kwargs) -> typing.Any: ...
+        def query_json(self, query: str, *args, **kwargs) -> str: ...
+        def query_one_json(self, query: str, *args, **kwargs) -> str: ...
+        def execute(self, query: str) -> None: ...
+
+These base classes should be implemented by respective classes:
+
+.. code-block:: python
+
+    class AsyncIOTransaction(AsyncExecutor): ...
+    class AsyncIOConnection(AsyncExecutor): ...
+    class AsyncIOPool(AsyncExecutor): ...
+    class Transaction(Executor): ...
+    class Connection(Executor): ...
+    class Pool(Executor): ...
+
+
+Exceptions API
+``````````````
+
+Error hierarchy is amended by introducing ``TransientError``, ``NetworkError``
+and ``EarlyNetworkError`` with the following relationships:
+
+.. code-block:: python
+
+    class TransientError(TransactionError): ...
+    class TransactionDeadlockError(TransientError): ...
+    class TransactionSerializationError(TransientError): ...
+    class NetworkError(EdgeDBError): ...
+    class EarlyNetworkError(NetworkError): ...
+
+All network error within connection should be converted into
+``EarlyNetworkError`` or ``NetworkError``. Former is used in context
+where we catch network error before sending a request.
+
+
+Future Work
+===========
+
+Transaction on Specific Connection
+----------------------------------
+
+Do we want and how ``connection.retry()`` should work?
+
+a. It may only retry on the same connection and fail on disconnect
+b. It may reconnect and replace underlying socket in the connection
+   object and retry
+c. We may only allow transactions on connection pools
```

Comment:
I'm not sure I understand any of these options. Is it about handling a network disconnect during an active transaction? FWIW, to me both (b) and (c) sound outright impractical. I suggest either expanding this section to give the reader more context or just remove it.

Reply:
Actually, (c) is what current RFC specifies. I think it's good enough until we have a clear use-case for having `Connection.retry`. If it non-concurrent ad-hoc script. I'd say connection pool with `min_size = 1` is fine. If it's for own connection handling then `try_transaction` is what they should use.